File: frontend/lib/models/reminder.dart
class Reminder {
  final String id;
  final String? customerId;
  final String? batteryId;
  final String customerName;
  final String mobileNumber;
  final String? batteryModel;
  final String? batterySerial;
  final String? batteryType;
  final String reminderType; // 'WATER_CHECK', 'SERVICE', 'WARRANTY_EXPIRY'
  final String reminderDate;
  final String? warrantyExpiry;
  final String reminderStatus; // 'UPCOMING', 'DUE', 'OVERDUE', 'COMPLETED', 'EXPIRED'
  final bool messageSent;
  final String? sentAt;
  final bool isCompleted;
  final bool isArchived;
  final String? notes;
  final String? whatsappTemplate;
  final String whatsappDeliveryStatus;
  final String? whatsappMessageId;
  final String createdAt;
  final String? updatedAt;
  final String? reminderCategory;
  final String? linkedPaymentId;
  final int recurringIntervalDays;
  final bool stopWhenSettled;

  Reminder({
    required this.id,
    this.customerId,
    this.batteryId,
    required this.customerName,
    required this.mobileNumber,
    this.batteryModel,
    this.batterySerial,
    this.batteryType,
    required this.reminderType,
    required this.reminderDate,
    this.warrantyExpiry,
    required this.reminderStatus,
    required this.messageSent,
    this.sentAt,
    required this.isCompleted,
    required this.isArchived,
    this.notes,
    this.whatsappTemplate,
    required this.whatsappDeliveryStatus,
    this.whatsappMessageId,
    required this.createdAt,
    this.updatedAt,
    this.reminderCategory = 'BATTERY',
    this.linkedPaymentId,
    this.recurringIntervalDays = 7,
    this.stopWhenSettled = true,
  });

  factory Reminder.fromJson(Map<String, dynamic> json) {
    return Reminder(
      id: json['id'] as String,
      customerId: json['customer_id'] as String?,
      batteryId: json['battery_id'] as String?,
      customerName: json['customer_name'] as String? ?? '',
      mobileNumber: json['mobile_number'] as String? ?? '',
      batteryModel: json['battery_model'] as String?,
      batterySerial: json['battery_serial'] as String?,
      batteryType: json['battery_type'] as String?,
      reminderType: json['reminder_type'] as String? ?? '',
      reminderDate: json['reminder_date'] as String? ?? '',
      warrantyExpiry: json['warranty_expiry'] as String?,
      reminderStatus: json['reminder_status'] as String? ?? 'UPCOMING',
      messageSent: json['message_sent'] as bool? ?? false,
      sentAt: json['sent_at'] as String?,
      isCompleted: json['is_completed'] as bool? ?? false,
      isArchived: json['is_archived'] as bool? ?? false,
      notes: json['notes'] as String?,
      whatsappTemplate: json['whatsapp_template'] as String?,
      whatsappDeliveryStatus: json['whatsapp_delivery_status'] as String? ?? 'PENDING',
      whatsappMessageId: json['whatsapp_message_id'] as String?,
      createdAt: json['created_at'] as String? ?? '',
      updatedAt: json['updated_at'] as String?,
      reminderCategory: json['reminder_category'] as String? ?? 'BATTERY',
      linkedPaymentId: json['linked_payment_id'] as String?,
      recurringIntervalDays: json['recurring_interval_days'] as int? ?? 7,
      stopWhenSettled: json['stop_when_settled'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'customer_id': customerId,
      'battery_id': batteryId,
      'customer_name': customerName,
      'mobile_number': mobileNumber,
      'battery_model': batteryModel,
      'battery_serial': batterySerial,
      'battery_type': batteryType,
      'reminder_type': reminderType,
      'reminder_date': reminderDate,
      'warranty_expiry': warrantyExpiry,
      'reminder_status': reminderStatus,
      'message_sent': messageSent,
      'sent_at': sentAt,
      'is_completed': isCompleted,
      'is_archived': isArchived,
      'notes': notes,
      'whatsapp_template': whatsappTemplate,
      'whatsapp_delivery_status': whatsappDeliveryStatus,
      'whatsapp_message_id': whatsappMessageId,
      'created_at': createdAt,
      'updated_at': updatedAt,
      'reminder_category': reminderCategory,
      'linked_payment_id': linkedPaymentId,
      'recurring_interval_days': recurringIntervalDays,
      'stop_when_settled': stopWhenSettled,
    };
  }

  Reminder copyWith({
    String? id,
    String? customerId,
    String? batteryId,
    String? customerName,
    String? mobileNumber,
    String? batteryModel,
    String? batterySerial,
    String? batteryType,
    String? reminderType,
    String? reminderDate,
    String? warrantyExpiry,
    String? reminderStatus,
    bool? messageSent,
    String? sentAt,
    bool? isCompleted,
    bool? isArchived,
    String? notes,
    String? whatsappTemplate,
    String? whatsappDeliveryStatus,
    String? whatsappMessageId,
    String? createdAt,
    String? updatedAt,
    String? reminderCategory,
    String? linkedPaymentId,
    int? recurringIntervalDays,
    bool? stopWhenSettled,
  }) {
    return Reminder(
      id: id ?? this.id,
      customerId: customerId ?? this.customerId,
      batteryId: batteryId ?? this.batteryId,
      customerName: customerName ?? this.customerName,
      mobileNumber: mobileNumber ?? this.mobileNumber,
      batteryModel: batteryModel ?? this.batteryModel,
      batterySerial: batterySerial ?? this.batterySerial,
      batteryType: batteryType ?? this.batteryType,
      reminderType: reminderType ?? this.reminderType,
      reminderDate: reminderDate ?? this.reminderDate,
      warrantyExpiry: warrantyExpiry ?? this.warrantyExpiry,
      reminderStatus: reminderStatus ?? this.reminderStatus,
      messageSent: messageSent ?? this.messageSent,
      sentAt: sentAt ?? this.sentAt,
      isCompleted: isCompleted ?? this.isCompleted,
      isArchived: isArchived ?? this.isArchived,
      notes: notes ?? this.notes,
      whatsappTemplate: whatsappTemplate ?? this.whatsappTemplate,
      whatsappDeliveryStatus: whatsappDeliveryStatus ?? this.whatsappDeliveryStatus,
      whatsappMessageId: whatsappMessageId ?? this.whatsappMessageId,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      reminderCategory: reminderCategory ?? this.reminderCategory,
      linkedPaymentId: linkedPaymentId ?? this.linkedPaymentId,
      recurringIntervalDays: recurringIntervalDays ?? this.recurringIntervalDays,
      stopWhenSettled: stopWhenSettled ?? this.stopWhenSettled,
    );
  }
}

class ReminderStats {
  final int todayFollowups;
  final int overdueCount;
  final int upcomingExpiry;
  final int waterChecksDue;
  final int pendingService;
  final int completed;

  ReminderStats({
    required this.todayFollowups,
    required this.overdueCount,
    required this.upcomingExpiry,
    required this.waterChecksDue,
    required this.pendingService,
    required this.completed,
  });

  factory ReminderStats.fromJson(Map<String, dynamic> json) {
    return ReminderStats(
      todayFollowups: json['today_followups'] as int? ?? 0,
      overdueCount: json['overdue_count'] as int? ?? 0,
      upcomingExpiry: json['upcoming_expiry'] as int? ?? 0,
      waterChecksDue: json['water_checks_due'] as int? ?? 0,
      pendingService: json['pending_service'] as int? ?? 0,
      completed: json['completed'] as int? ?? 0,
    );
  }
}

File: backend/app/models/reminder.py
from datetime import date, datetime
from typing import Optional
from pydantic import BaseModel, field_validator


# ---------------------------------------------------------------------------
# Request models
# ---------------------------------------------------------------------------

class ReminderCreate(BaseModel):
    model_config = {
        "protected_namespaces": ()
    }

    customer_id: Optional[str] = None
    battery_id: Optional[str] = None
    customer_name: str
    mobile_number: str
    battery_model: Optional[str] = None
    battery_serial: Optional[str] = None
    battery_type: Optional[str] = None
    reminder_type: str  # 'WATER_CHECK', 'SERVICE', 'WARRANTY_EXPIRY', 'UDHARI'
    reminder_date: date
    warranty_expiry: Optional[date] = None
    notes: Optional[str] = None
    whatsapp_template: Optional[str] = None
    reminder_category: Optional[str] = 'BATTERY'
    linked_payment_id: Optional[str] = None
    recurring_interval_days: Optional[int] = 7
    stop_when_settled: Optional[bool] = True

    @field_validator("reminder_type")
    @classmethod
    def validate_reminder_type(cls, v: str) -> str:
        allowed = {"WATER_CHECK", "SERVICE", "WARRANTY_EXPIRY", "UDHARI", "UDHARI_RECOVERY"}
        upper = v.strip().upper()
        if upper not in allowed:
            raise ValueError(f"reminder_type must be one of: {allowed}")
        return upper


class ReminderUpdate(BaseModel):
    model_config = {
        "protected_namespaces": ()
    }

    customer_name: Optional[str] = None
    mobile_number: Optional[str] = None
    battery_model: Optional[str] = None
    battery_serial: Optional[str] = None
    battery_type: Optional[str] = None
    reminder_date: Optional[date] = None
    warranty_expiry: Optional[date] = None
    is_completed: Optional[bool] = None
    notes: Optional[str] = None
    message_sent: Optional[bool] = None
    sent_at: Optional[datetime] = None
    whatsapp_template: Optional[str] = None
    whatsapp_delivery_status: Optional[str] = None
    whatsapp_message_id: Optional[str] = None


# ---------------------------------------------------------------------------
# Response model
# ---------------------------------------------------------------------------

class ReminderResponse(BaseModel):
    model_config = {
        "protected_namespaces": ()
    }

    id: str
    customer_id: Optional[str]
    battery_id: Optional[str]
    customer_name: str
    mobile_number: str
    battery_model: Optional[str]
    battery_serial: Optional[str]
    battery_type: Optional[str]
    reminder_type: str
    reminder_date: str
    warranty_expiry: Optional[str]
    reminder_status: str
    message_sent: bool
    sent_at: Optional[str]
    is_completed: bool
    is_archived: bool
    notes: Optional[str]
    whatsapp_template: Optional[str]
    whatsapp_delivery_status: str
    whatsapp_message_id: Optional[str]
    created_at: str
    updated_at: Optional[str] = None
    reminder_category: Optional[str] = 'BATTERY'
    linked_payment_id: Optional[str] = None
    recurring_interval_days: Optional[int] = 7
    stop_when_settled: Optional[bool] = True

    @classmethod
    def from_row(cls, row: dict) -> "ReminderResponse":
        # Calculate reminder_status dynamically based on date rules
        is_completed = row.get("is_completed", False)
        reminder_date_str = str(row["reminder_date"])
        warranty_expiry_str = str(row["warranty_expiry"]) if row.get("warranty_expiry") else None

        status_val = "UPCOMING"
        if is_completed:
            status_val = "COMPLETED"
        else:
            today = date.today()
            rem_date = date.fromisoformat(reminder_date_str)
            if rem_date > today:
                status_val = "UPCOMING"
            elif rem_date == today:
                status_val = "DUE"
            else:  # rem_date < today
                if warranty_expiry_str:
                    expiry = date.fromisoformat(warranty_expiry_str)
                    if expiry < today:
                        status_val = "EXPIRED"
                    else:
                        status_val = "OVERDUE"
                else:
                    status_val = "OVERDUE"

        return cls(
            id=str(row["id"]),
            customer_id=str(row["customer_id"]) if row.get("customer_id") else None,
            battery_id=str(row["battery_id"]) if row.get("battery_id") else None,
            customer_name=row["customer_name"],
            mobile_number=row["mobile_number"],
            battery_model=row.get("battery_model"),
            battery_serial=row.get("battery_serial"),
            battery_type=row.get("battery_type"),
            reminder_type=row["reminder_type"],
            reminder_date=reminder_date_str,
            warranty_expiry=warranty_expiry_str,
            reminder_status=status_val,
            message_sent=row.get("message_sent", False),
            sent_at=str(row["sent_at"]) if row.get("sent_at") else None,
            is_completed=is_completed,
            is_archived=row.get("is_archived", False),
            notes=row.get("notes"),
            whatsapp_template=row.get("whatsapp_template"),
            whatsapp_delivery_status=row.get("whatsapp_delivery_status", "PENDING"),
            whatsapp_message_id=row.get("whatsapp_message_id"),
            created_at=str(row.get("created_at", "")),
            updated_at=str(row["updated_at"]) if row.get("updated_at") else None,
            reminder_category=row.get("reminder_category", "BATTERY"),
            linked_payment_id=str(row["linked_payment_id"]) if row.get("linked_payment_id") else None,
            recurring_interval_days=row.get("recurring_interval_days", 7),
            stop_when_settled=row.get("stop_when_settled", True),
        )


class ReminderListResponse(BaseModel):
    data: list[ReminderResponse]
    total: int
    page: int
    limit: int


class ReminderStatsResponse(BaseModel):
    today_followups: int
    upcoming_expiry: int
    water_checks_due: int
    pending_service: int
    completed: int
